Add `FrequencyReachAdditiveEffect`

PabloRoque opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-labs/pymc-marketing/releases -->

## Description
<!--- Describe your changes in detail -->
Add Frequency/Reach support via the `MuEffect` interface. 
Similar, but slightly different (see below) to the implementation described [here](https://developers.google.com/meridian/docs/advanced-modeling/reach-frequency)

Notes:

~~1) The formulation is not strictly equivalent. Notice that our `frequency_sat: Tensorvariable` already applies the multiplicative `Beta` coefficient. If we want a fully equivalent implementation we need to call , but then we cannot rely on `.apply` to do dim_handling.~~

1) Introduced `HillShapeSaturation` to make the formulation equivalent

2) This is a very drafty feature. No support for optimization whatsoever.

## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [ ] Closes #
- [ ] Related to #

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [ ] Checked that [the pre-commit linting/style checks pass](https://www.pymc-marketing.io/en/latest/contributing/index.html). Feel free to comment [`pre-commit.ci autofix` to auto-fix](https://pre-commit.ci/#configuration-autofix_prs).
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks) using [numpydoc format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->


<!-- readthedocs-preview pymc-marketing start -->
----
📚 Documentation preview 📚: https://pymc-marketing--1968.org.readthedocs.build/en/1968/

<!-- readthedocs-preview pymc-marketing end -->

PabloRoque committed:
Add FrequencyReachAdditiveEffect

williambdean commented on pymc_marketing/mmm/frequency_reach_effect.py:
Shall we change additive_effect.py into additive_effects/ submodule and has this as a file?

PabloRoque committed:
Remove unused adstock_first arg

PabloRoque committed:
Add HillShapeSaturation

PabloRoque committed:
Freq-reach effect on HillShape

PabloRoque committed:
Add HillShapeSaturation to __init__

PabloRoque committed:
Introduce additive_effect submodule

PabloRoque committed:
Add effects to __init__

juanitorduz commented:
Amazing!

TeemuSailynoja commented on pymc_marketing/mmm/additive_effect/frequency_reach_effect.py:
```diff
@@ -0,0 +1,416 @@
+* Duplicate rows for the same (date, dims, channel) are aggregated (mean).
+* If channels present in the DataFrame are not a subset of the MMM channel
+  coordinate, a ValueError is raised.
+
```
Why is this required? I would assume the opposite. Isn't reach + frequency an alternative to using impressions or spend, and having both in the model would effectively double count the contribution of the involved channels?
The estimated contributions are also separate from the `channel_contribution` deterministic in the parent MMM, so this could lead to confusion.

PabloRoque replied:
You're right — R&F is an *alternative* to impressions/spend for those channels, not an addition. Using both would double-count their contribution.\n\nMeridian enforces this the same way: `InputData._validate_media_channels` raises a `ValueError` if any channel name appears in more than one of `media_channel`, `rf_channel`, etc.\n\nWe've added the same guard in `create_data` — it now raises with a clear message if any R&F channel name overlaps with `mmm.channel_columns`. The docstring has also been updated to make the alternative-not-additive semantics explicit.

TeemuSailynoja commented on pymc_marketing/mmm/additive_effect/frequency_reach_effect.py:
```diff
@@ -0,0 +1,416 @@
+            mu=0.0,
+            sigma=1.0,
+            dims=(self.channel_coord_name,),
+        )
```
perhaps expose to user as an optional argument of the Prior class, and have this as a default value.

PabloRoque replied:
Good call. Changed the default to `HalfNormal(sigma=1)` — semantically correct for an R&F scaling coefficient (contributions should be non-negative). Users can still pass any `Prior` they like at construction time, e.g. `beta_prior=Prior("Gamma", alpha=2, beta=1)`.

TeemuSailynoja reviewed (changes requested):
The order of adstock and saturation is reversed in the documentation.
Should we reference [the source paper](https://research.google/pubs/bayesian-hierarchical-media-mix-model-incorporating-reach-and-frequency-data/) instead of Meridian documentation?

TeemuSailynoja commented:
@PabloRoque this is cool! Need any help? @juanitorduz wants this 🤣

PabloRoque committed:
Add FrequencyReachAdditiveEffect with HillShapeSaturation and budget optimizer support

- Add HillShapeSaturation to saturation.py with serialization registration
  and correct dim kwarg in function signature; restore RootSaturation dropped
  during rebase conflict resolution
- Convert additive_effect.py to additive_effect/ subpackage; add
  frequency_reach_effect.py implementing the Meridian-style R&F pipeline:
  - Hardcoded RF_CHANNEL_COORD='rf_channel' disjoint from standard 'channel'
  - Configurable beta_prior (default Normal(0,1))
  - cost_per_unit and assumed_frequency fields for optimizer support
  - create_data / create_effect / set_data following MuEffect protocol
  - build_rf_optimization_tensors() for budget optimizer graph injection
  - _build_raw_arrays() helper accepting optional model param for set_data
- Extend BudgetOptimizer with _replace_rf_data_by_optimization_variable()
  private method; called from model_post_init as a no-op when no R&F effects
  are present
- Export FrequencyReachAdditiveEffect and RF_CHANNEL_COORD from
  pymc_marketing.mmm.__init__
- Add test suite (34 tests) covering HillShapeSaturation, validation,
  properties, model integration, set_data, optimizer tensors, and noop path

PabloRoque committed:
fix(freq-reach): migrate to pmd.Data/pmd.Deterministic XTensor backend

- Replace pm.Data with pmd.Data in create_data; returns XTensorVariable
  directly, eliminating the manual ptx.as_xtensor re-wrap workaround
- Remove ptx.as_xtensor(model[...], dims=...) calls from create_effect
- Replace pm.Deterministic with pmd.Deterministic throughout create_effect
- Replace pm.set_data with direct .set_value() on XTensorSharedVariable
  in set_data; pm.set_data does not handle XTensorSharedVariable and
  raises a float64/float32 precision error regardless of input dtype
- Add import pymc.dims as pmd

PabloRoque commented on pymc_marketing/mmm/additive_effect/frequency_reach_effect.py:
```diff
@@ -0,0 +1,416 @@
+    saturation : SaturationTransformation
+        Saturation transformation applied after adstock.
+    adstock : AdstockTransformation
+        Adstock transformation applied first to the effective exposure signal.
```
Thanks for catching this — you're right that the code is correct. To confirm the intended order: Meridian applies Hill saturation to frequency first, then adstock to the effective exposure signal (reach x saturated_frequency). This is consistent with the Meridian source code in adstock_hill.py, where AdstockTransformer.forward explicitly notes: the Hill function can be applied before or after Adstock, and Hill-first is the intended path for R&F channels. The current docstrings already reflect the correct order (saturation then adstock), so no code change is needed here.

PabloRoque committed:
feat(freq-reach): validate R&F channels are disjoint from channel_columns

R&F data is an alternative representation for those channels, not an
addition. Mixing the same channel in both channel_columns and
df_frequency_reach would double-count its contribution.

- Raise ValueError in create_data when any R&F channel name appears in
  mmm.channel_columns; access via getattr so the Model protocol is not
  polluted (MVITS and other non-MMM protocol implementors are unaffected)
- Error message mirrors Meridian InputData._validate_media_channels,
  which raises on the same condition
- Add test_create_data_raises_on_channel_overlap in TestFrequencyReachModelIntegration
- Update module docstring to explain the disjointness requirement and the
  alternative-not-additive semantics
- Remove AI-style separator comment blocks from test file

PabloRoque committed:
fix(freq-reach): change beta_prior default to HalfNormal(sigma=1)
